nbodykit/base/catalogmesh.py

```diff
@@ -38,66 +39,69 @@ class CatalogMesh(MeshSource, CatalogSource):
         is ``Position``
     """
     logger = logging.getLogger('CatalogMesh')
+
     def __repr__(self):
-        return "(%s as CatalogMesh)" % repr(self.source)
+        return "(%s as CatalogMesh)" % repr(self.base)
 
-    # intended to be used by CatalogSource internally
-    def __init__(self, source, BoxSize, Nmesh, dtype, weight,
-                    value, selection, position='Position'):
+    def __new__(cls, source, BoxSize, Nmesh, dtype, weight,
+                    value, selection, position='Position', interlaced=False,
+                    compensated=False, window='cic', **kwargs):
 
-        # ensure self.comm is set, though usually already set by the child.
-        self.comm = source.comm
-        self.source = source
-        self.position = position
-        self.selection = selection
-        self.weight = weight
-        self.value = value
+        assert isinstance(source, CatalogSourceBase)
 
-        self.attrs.update(source.attrs)
+        # view the input source (a CatalogSource) as the desired class
+        obj = source._view(cls)
 
-        # this will override BoxSize and Nmesh carried from the source, if there is any!
-        MeshSource.__init__(self, BoxSize=BoxSize, Nmesh=Nmesh, dtype=dtype, comm=source.comm)
-        CatalogSource.__init__(self, comm=source.comm)
+        # copy over the necessary meta-data to attrs
+        obj.attrs['BoxSize'] = BoxSize
+        obj.attrs['Nmesh'] = Nmesh
+        obj.attrs['interlaced'] = interlaced
+        obj.attrs['compensated'] = compensated
+        obj.attrs['window'] = window
 
-        # copy over the overrides
-        self._overrides.update(self.source._overrides)
+        # store others as straight attributes
+        obj.dtype = dtype
+        obj.weight = weight
+        obj.value = value
+        obj.selection = selection
+        obj.position = position
 
-        self.attrs['position'] = self.position
-        self.attrs['selection'] = self.selection
-        self.attrs['weight'] = self.weight
-        self.attrs['value'] = self.value
-        self.attrs['compensated'] = True
-        self.attrs['interlaced'] = False
-        self.attrs['window'] = 'cic'
+        # add in the Mesh Source attributes
+        MeshSource.__finalize__(obj, obj)
 
-    @property
-    def size(self):
-        """
-        The number of local particles.
-        """
-        return self.source.size
+        return obj
 
-    @property
-    def hardcolumns (self):
-        """
-        The names of the hard-coded columns in the source.
-        """
-        return self.source.hardcolumns
+    def __init__(self, *args,  **kwargs):
+        pass
 
-    def get_hardcolumn(self, col):
-        """
-        Return a hard-coded column
-        """
-        return self.source.get_hardcolumn(col)
+    def __finalize__(self, obj):
```

Comment:
Why not using  __init__? Another possibilty is to use `setstate`

I think if __new__ exists you have to explicitly trigger __init__ anyways.. (I may be wrong).

Reply:
``obj`` here can be a CatalogSource or a CatalogMesh, so I'm not sure ``__init__`` would work. I guess this is sort of a ``setstate``.

``CatalogMesh.__finalize__`` will get called in different situations: 

1. when slicing a CatalogMesh to get a subset

https://github.com/bccp/nbodykit/blob/884f57ca748149fe7871138b01285d4c65fdea88/nbodykit/base/catalog.py#L434

2. when initializing a CatalogMesh from a CatalogSource

https://github.com/bccp/nbodykit/blob/884f57ca748149fe7871138b01285d4c65fdea88/nbodykit/base/catalogmesh.py#L53

Reply:
I guess what makes me confused is the function is defined by the occasion it is called, but not by its purpose -- thus it starts to feel like we are building a framework.

If I read the function, the purpose of the function is to copy Python attributes? would it work if we simply copy the `__dict__` of the object instead?

Reply:
The logic is the same as numpy's ``__array_finalize__``. In this case, copying ``__dict__`` won't work because when you are going from CatalogSource to CatalogMesh, ``self`` in ``__finalize__(self, obj)`` is a bare object, and you need to add the default CatalogMesh attributes to it.